File: packages/gavio-js/src/interceptors/pii/scanner.ts
/** PiiScanner interface and ScannerRegistry. */

import type { ScanContext } from './context.js'
import type { PiiMatch } from './match.js'

/**
 * Detects one class of PII entity within text.
 *
 * Scanners are tiered: tier 1 = regex, tier 2 = NER/ML, tier 3 = LLM. Lower
 * tiers run first so cheap deterministic matches are found before expensive
 * ones. v0.1.0 ships only tier-1 regex scanners.
 */
export interface PiiScanner {
  /** e.g. 'EMAIL', 'IBAN', 'BSN'. */
  readonly entityType: string
  /** default: 1 */
  readonly tier?: 1 | 2 | 3
  scan(text: string, ctx: ScanContext): PiiMatch[] | Promise<PiiMatch[]>
  /** default: 1.0 */
  readonly confidence?: number
  supportsLanguage?(lang: string): boolean
  supportsLocale?(locale: string): boolean
}

export function scannerTier(s: PiiScanner): number {
  return s.tier ?? 1
}

/** Registry of scanners, discoverable by entity type at runtime. */
export class ScannerRegistry {
  private scanners: PiiScanner[] = []

  constructor(scanners?: PiiScanner[]) {
    for (const s of scanners ?? []) this.register(s)
  }

  register(scanner: PiiScanner): this {
    this.scanners.push(scanner)
    return this
  }

  /** Return scanners sorted by tier (lowest first). */
  all(): PiiScanner[] {
    return [...this.scanners].sort((a, b) => scannerTier(a) - scannerTier(b))
  }

  byEntityType(entityType: string): PiiScanner[] {
    return this.scanners.filter((s) => s.entityType === entityType)
  }

  get size(): number {
    return this.scanners.length
  }
}

File: packages/gavio-py/gavio/interceptors/pii/scanner.py
"""PiiScanner abstract base class and ScannerRegistry."""

from __future__ import annotations

from abc import ABC, abstractmethod

from .context import ScanContext
from .match import PiiMatch


class PiiScanner(ABC):
    """Detects one class of PII entity within text.

    Scanners are tiered: tier 1 = regex, tier 2 = NER/ML, tier 3 = LLM. Lower
    tiers run first so cheap deterministic matches are found before expensive
    ones. v0.1.0 ships only tier-1 regex scanners.
    """

    @property
    @abstractmethod
    def entity_type(self) -> str:
        """e.g. 'EMAIL', 'IBAN', 'BSN'."""
        ...

    @property
    def tier(self) -> int:
        return 1

    @abstractmethod
    def scan(self, text: str, ctx: ScanContext) -> list[PiiMatch]:
        ...

    def confidence(self) -> float:
        return 1.0

    def supports_language(self, lang: str) -> bool:
        return True

    def supports_locale(self, locale: str) -> bool:
        return True


class ScannerRegistry:
    """Registry of scanners, discoverable by entity type at runtime."""

    def __init__(self, scanners: list[PiiScanner] | None = None) -> None:
        self._scanners: list[PiiScanner] = []
        for scanner in scanners or []:
            self.register(scanner)

    def register(self, scanner: PiiScanner) -> ScannerRegistry:
        self._scanners.append(scanner)
        return self

    def scanners(self) -> list[PiiScanner]:
        """Return scanners sorted by tier (lowest first)."""
        return sorted(self._scanners, key=lambda s: s.tier)

    def by_entity_type(self, entity_type: str) -> list[PiiScanner]:
        return [s for s in self._scanners if s.entity_type == entity_type]

    def __len__(self) -> int:
        return len(self._scanners)
